File: src/main/java/net/b0n541/pmcts/game/tictactoe/TicTacToePlayer.java
package net.b0n541.pmcts.game.tictactoe;

import net.b0n541.pmcts.mcts.GamePlayer;
import net.b0n541.pmcts.mcts.MonteCarloTreeSearch;
import net.b0n541.pmcts.mcts.Tree;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;

public class TicTacToePlayer implements GamePlayer {

    private static final Logger LOG = LoggerFactory.getLogger(TicTacToePlayer.class);

    public final PlayerSymbol playerSymbol;
    private TicTacToeGameState gameState;

    public TicTacToePlayer(final PlayerSymbol playerSymbol, final PlayerSymbol firstPlayer) {
        this.playerSymbol = playerSymbol;
        gameState = new TicTacToeGameState(firstPlayer);
    }

    public TicTacToeMove play() {
        final Tree tree = new Tree(gameState);

        play(tree, Duration.ofSeconds(1));
        //play(tree, 200);

        //tree.printDigraph();

        return (TicTacToeMove) tree.getBestMove();
    }

    private static void play(final Tree tree, final int maxRounds) {
        long rounds = 0;
        do {
            MonteCarloTreeSearch.run(tree, 1);
            rounds++;
        } while (rounds < maxRounds);

        LOG.info("Tree search finished. Rounds: {}, Tree size: {}", rounds, tree.getSize());
    }

    private static void play(final Tree tree, final Duration duration) {
        final long finishTime = System.nanoTime() + duration.toNanos();
        long rounds = 0;
        do {
            MonteCarloTreeSearch.run(tree, 1);
            rounds++;
        } while (System.nanoTime() < finishTime);

        LOG.info("Tree search finished. Rounds: {}, Tree size: {}", rounds, tree.getSize());
    }

    public void addMove(final TicTacToeMove move) {
        gameState = gameState.addMove(move);
    }

    @Override
    public String toShortString() {
        return playerSymbol.toString();
    }
}

File: src/main/kotlin/net/b0n541/ai/game/tictactoe/TicTacToePlayer.kt
package net.b0n541.ai.game.tictactoe

import net.b0n541.ai.mcts.GameMove
import net.b0n541.ai.mcts.GamePlayer
import net.b0n541.ai.mcts.GameState
import net.b0n541.ai.mcts.MonteCarloTreeSearch.run
import net.b0n541.ai.mcts.Tree
import org.slf4j.LoggerFactory
import java.time.Duration

open class TicTacToePlayer(val playerSymbol: PlayerSymbol, firstPlayer: PlayerSymbol) : GamePlayer<TicTacToeMove> {

    private var gameState: TicTacToeGameState

    open fun play(): TicTacToeMove {
        val tree = Tree(gameState as GameState<GameMove>)
        play(tree, Duration.ofSeconds(1))

        //play(tree, 200)
        //tree.printDigraph()

        return tree.bestMove as TicTacToeMove
    }

    override fun addMove(move: TicTacToeMove) {
        gameState = gameState.addMove(move)
    }

    override fun toShortString(): String {
        return playerSymbol.toString()
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(TicTacToePlayer::class.java)
        private fun play(tree: Tree, maxRounds: Int) {
            var rounds: Long = 0
            do {
                run(tree, 1)
                rounds++
            } while (rounds < maxRounds)
            LOG.info("Tree search finished. Rounds: {}, Tree size: {}", rounds, tree.size)
        }

        private fun play(tree: Tree, duration: Duration) {
            val finishTime = System.nanoTime() + duration.toNanos()
            var rounds: Long = 0
            do {
                run(tree, 1)
                rounds++
            } while (System.nanoTime() < finishTime)
            LOG.info("Tree search finished. Rounds: {}, Tree size: {}", rounds, tree.size)
        }
    }

    init {
        gameState = TicTacToeGameState(firstPlayer)
    }
}